lib/srv/app/aws/endpoints.go

```diff
@@ -128,6 +150,122 @@ func resolveEndpointByXForwardedHost(r *http.Request, headerKey string) (*endpoi
+	sessCtx, err := common.GetSessionContext(r)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	awsAuthHeader, err := awsutils.ParseSigV4(r.Header.Get(awsutils.AuthorizationHeader))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	credentials, err := s.CredentialsGetter.Get(s.closeContext, awsutils.GetCredentialsRequest{
+		Provider:    s.Session,
+		Expiry:      sessCtx.Identity.Expires,
+		SessionName: sessCtx.Identity.Username,
+		RoleARN:     sessCtx.Identity.RouteToApp.AWSRoleARN,
+		ExternalID:  sessCtx.App.GetAWSExternalID(),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key := resolveTimestreamEndpointCacheKey{
+		credentials:               credentials,
+		region:                    awsAuthHeader.Region,
+		isTimemstreamWriteRequest: isTimestreamWriteRequest(r),
+	}
+	re, err := utils.FnCacheGet(s.closeContext, s.cache, key, func(ctx context.Context) (*endpoints.ResolvedEndpoint, error) {
+		if key.isTimemstreamWriteRequest {
+			return resolveTimestreamWriteEndpoint(ctx, key.credentials, key.region)
+		}
+		return resolveTimestreamQueryEndpoint(ctx, key.credentials, key.region)
+	})
+	return re, trace.Wrap(err)
+}
+
+type resolveTimestreamEndpointCacheKey struct {
+	credentials               *credentials.Credentials
+	region                    string
+	isTimemstreamWriteRequest bool
+}
+
+func isTimestreamWriteRequest(r *http.Request) bool {
+	switch strings.TrimPrefix(r.Header.Get("X-Amz-Target"), timestreamOpPrefix) {
+	case "CreateDatabase", "CreateTable",
+		"DeleteDatabase", "DeleteTable",
+		"DescribeDatabase", "DescribeTable",
+		"ListDatabases", "ListTables",
+		"UpdateDatabase", "UpdateTable",
+		"WriteRecords":
+		return true
+	// Note that both timestream-query and timestream-write support these tag operations.
+	// For now, we assume they are used for timestream-query.
+	case "ListTagsForResource", "TagResource", "UntagResource":
+		return false
+	default:
+		return false
+	}
+}
+
+func resolveTimestreamWriteEndpoint(ctx context.Context, credentials *credentials.Credentials, region string) (*endpoints.ResolvedEndpoint, error) {
+	session, err := session.NewSession(aws.NewConfig().WithCredentials(credentials).WithRegion(region))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	output, err := timestreamwrite.New(session).DescribeEndpointsWithContext(ctx, &timestreamwrite.DescribeEndpointsInput{})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if output == nil || len(output.Endpoints) == 0 {
+		return nil, trace.NotFound("no timestream-write endpoints found")
+	}
+	return &endpoints.ResolvedEndpoint{
+		URL:           "https://" + aws.StringValue(output.Endpoints[0].Address),
+		SigningRegion: region,
+		SigningName:   "timestream",
+	}, nil
+}
+func resolveTimestreamQueryEndpoint(ctx context.Context, credentials *credentials.Credentials, region string) (*endpoints.ResolvedEndpoint, error) {
```

Comment:
The duplication between `resolveTimestreamQueryEndpoint` and `resolveTimestreamWriteEndpoint` is unfortunate. Can we share this logic somehow?